libraries/automaton/automaton.lua

```diff
@@ -0,0 +1,168 @@
+local list = require('list')
+local items = require('items')
+local packet = require('packet')
+local player = require('player')
+local resources = require('resources')
+local client = require('shared.client')
+
+local data, ftype = client.new('automaton_service')
+
+local res_items = resources.items
+local assembly_offset, attachment_offset = 0x2000, 0x2100
+
+ftype.fields.head = {
+    get = function(data)
+        return resources.items[data.head_id + assembly_offset]
```

Comment:
Well the internal id is needed for packet construction, where the item id is not needed, the `remove_all` is the only function that actually takes advantage of this though

Reply:
Yeah that's fine, it does not rely on `data` having that internal ID.

Reply:
it would change this 
```lua
        for k, v in pairs(data.attachments) do
            pay_load.slots[k + 1] = v.id
        end
```
to
```lua
        for k, v in pairs(data.attachments) do
            pay_load.slots[k + 1] = v.id + attachment_offset
        end
```
in the `remove_all` function

Reply:
It would be `- attachment_offset`, not `+`, I think. But yes. Is this a problem?

> also currently the attachments have id, slot, and item as fields, if we do id = item.id that means we lose some information and do not gain any.

We only lose information we don't need. That, by itself, is beneficial, but we also gain performance in the read case, and we always prioritize the read over the write case. If an addon wants to display that info on-screen, for example, it needs to read that every frame, so for the head, frame and all attachments it always has to add the offset, every frame. Like this it only does it once when the new info comes in.

It speculates that the critical case will be reading data over writing data, which is a reasonable assumption to make in most cases.

Reply:
🤔 ok, but if we remove the offset well setting the value in the service, now both the service and library need to know the right offset value, currently only the library needs it.

Reply:
I don't see why it's an issue to have that value in both places, I'd just manually add `0x2000` or `0x2100`. I really would not like having it as a data variable, that way we'd expose another useless piece of data to the end user. If you are really want to have it in one place only, you can use a separately shared structure, which only the library consumes but does not expose further, but I still don't see the need for it. It's not like you're repeating dozens of lines of code, just a single fixed value.

Reply:
Actually, just thought if you really want it you can call it `data._attachment_offset`. I could live with that, since that remains accessible but it won't show up in `pairs`, and we can ignore it in the documentation. I'd still prefer having just the literal value there, though.

Reply:
Well wont the dev need the offset too? I mean if they happen to want to do something with the internal ids. Really want to nail this service and library so i can use it as the example for doing the blue mage equivalent

Reply:
The only reason I can see for the dev wanting that value is if they want to do something with injection or memory manipulation that the library doesn't provide. In that case it's either an extremely specialized use-case and not something a general purpose library should support, or a feature the library is currently be missing and it should be added to the library in that case.